File: runtime/src/cadence.ts
/**
 * When, as the hub and the app write it, and as this runtime's timer reads it.
 *
 * The app sends structure — a repeat, a time, a zone — so nothing is guessed.
 * The timer here speaks cron. This is the translation, both ways, and the
 * words the person is shown before they agree to anything.
 */

export interface Cadence {
  repeat: "once" | "daily" | "weekdays" | "weekly";
  /** "07:00", 24-hour, in `timezone`. */
  time: string;
  /** Weekly only: "mon" … "sun". */
  days?: string[];
  /** Once only: "2026-09-12". */
  date?: string;
  timezone: string;
}

const DAYS = ["sun", "mon", "tue", "wed", "thu", "fri", "sat"];

/** The cron line for a cadence. A one-off becomes a date, run once. */
export function cronOf(cadence: Cadence): string {
  const [hour, minute] = cadence.time.split(":").map((part) => Number(part));
  const at = `${minute ?? 0} ${hour ?? 0}`;
  switch (cadence.repeat) {
    case "daily":
      return `${at} * * *`;
    case "weekdays":
      return `${at} * * 1-5`;
    case "weekly":
      return `${at} * * ${(cadence.days ?? []).map((d) => DAYS.indexOf(d)).filter((d) => d >= 0).join(",")}`;
    case "once": {
      const [, month, day] = (cadence.date ?? "").split("-").map((part) => Number(part));
      return `${at} ${day} ${month} *`;
    }
  }
}

/**
 * The cadence a cron line means, when it is one the app can show: routines
 * set before the app could see them are carried across this way. Anything
 * cleverer than a time on some days of the week is left where it is.
 */
export function cadenceOf(cron: string, timezone: string): Cadence | undefined {
  const match = /^(\d{1,2}) (\d{1,2}) \* \* (\*|1-5|[0-6](?:,[0-6])*)$/.exec(cron.trim());
  if (!match) return undefined;
  const [, minute, hour, days] = match;
  const time = `${hour!.padStart(2, "0")}:${minute!.padStart(2, "0")}`;
  if (Number(hour) > 23 || Number(minute) > 59) return undefined;
  if (days === "*") return { repeat: "daily", time, timezone };
  if (days === "1-5") return { repeat: "weekdays", time, timezone };
  return { repeat: "weekly", time, days: [...new Set(days!.split(",").map((d) => DAYS[Number(d)]!))], timezone };
}

/** The cadence in plain words, for saying back before anything is set. */
export function describe(cadence: Cadence): string {
  const [hour, minute] = cadence.time.split(":").map((part) => Number(part));
  const h = hour ?? 0;
  const clock = `${h % 12 || 12}:${String(minute ?? 0).padStart(2, "0")} ${h < 12 ? "am" : "pm"}`;
  switch (cadence.repeat) {
    case "daily":
      return `every day at ${clock}`;
    case "weekdays":
      return `every weekday at ${clock}`;
    case "weekly":
      return `every ${(cadence.days ?? []).join(", ")} at ${clock}`;
    case "once":
      return `on ${cadence.date} at ${clock}`;
  }
}

/** Whether a cadence is one this runtime can hold. */
export function validCadence(cadence: Cadence): boolean {
  if (!/^([01]\d|2[0-3]):[0-5]\d$/.test(cadence.time)) return false;
  if (cadence.repeat === "weekly") return (cadence.days ?? []).every((d) => DAYS.includes(d)) && !!cadence.days?.length;
  if (cadence.repeat === "once") return /^\d{4}-\d{2}-\d{2}$/.test(cadence.date ?? "");
  return cadence.repeat === "daily" || cadence.repeat === "weekdays";
}

File: server/internal/conversations/cadence.go
package conversations

import (
	"fmt"
	"slices"
	"strings"
	"time"
	// The hub's image carries no zone files of its own; the list is small
	// enough to take along, and "Asia/Kolkata" must mean the same on every
	// machine.
	_ "time/tzdata"

	"github.com/Saieshwar5/cuckoo/server/internal/domain"
)

// Cadence is when a schedule runs, as structure rather than a sentence.
//
// A person picks a time and a repeat on their phone; nothing reads "7" and
// decides it meant the evening. The agent interprets what to do, never when.
type Cadence struct {
	// Repeat is once, daily, weekdays or weekly.
	Repeat string `json:"repeat"`
	// Time is the hour and minute, "07:00", on a 24-hour clock, in Timezone.
	Time string `json:"time"`
	// Days, for weekly only: "mon" … "sun".
	Days []string `json:"days,omitempty"`
	// Date, for once only: "2026-09-12".
	Date string `json:"date,omitempty"`
	// Timezone is where the clock is: "Asia/Kolkata".
	Timezone string `json:"timezone"`
}

// How often a schedule repeats. Nothing more often than daily on purpose:
// hourly is where a person's own schedule turns into the agent's spam.
const (
	RepeatOnce     = "once"
	RepeatDaily    = "daily"
	RepeatWeekdays = "weekdays"
	RepeatWeekly   = "weekly"
)

var weekdays = []string{"sun", "mon", "tue", "wed", "thu", "fri", "sat"}

// missGrace is how late a run may be before the app says it did not happen.
const missGrace = 10 * time.Minute

func invalidCadence(msg string) error {
	return domain.InvalidField("cadence", "invalid_cadence", msg)
}

// normalized checks a cadence and returns it in its one written form: days
// lower-case, deduplicated and in week order.
func (c Cadence) normalized() (Cadence, error) {
	switch c.Repeat {
	case RepeatOnce, RepeatDaily, RepeatWeekdays, RepeatWeekly:
	default:
		return Cadence{}, invalidCadence(`repeat must be "once", "daily", "weekdays" or "weekly".`)
	}
	if _, err := time.Parse("15:04", c.Time); err != nil || len(c.Time) != 5 {
		return Cadence{}, invalidCadence(`time must be an hour and minute on a 24-hour clock, like "07:00".`)
	}
	if c.Timezone == "" {
		return Cadence{}, invalidCadence(`timezone is needed, like "Asia/Kolkata".`)
	}
	if _, err := time.LoadLocation(c.Timezone); err != nil {
		return Cadence{}, invalidCadence(fmt.Sprintf("%q is not a time zone.", c.Timezone))
	}

	out := Cadence{Repeat: c.Repeat, Time: c.Time, Timezone: c.Timezone}
	if c.Repeat == RepeatWeekly {
		for _, d := range weekdays {
			for _, given := range c.Days {
				if strings.EqualFold(strings.TrimSpace(given), d) && !slices.Contains(out.Days, d) {
					out.Days = append(out.Days, d)
				}
			}
		}
		if len(out.Days) == 0 || len(out.Days) != len(uniqueLower(c.Days)) {
			return Cadence{}, invalidCadence(`a weekly schedule names its days, from "mon" to "sun".`)
		}
	} else if len(c.Days) > 0 {
		return Cadence{}, invalidCadence("only a weekly schedule names days.")
	}
	if c.Repeat == RepeatOnce {
		if _, err := time.Parse("2006-01-02", c.Date); err != nil {
			return Cadence{}, invalidCadence(`a one-off schedule has a date, like "2026-09-12".`)
		}
		out.Date = c.Date
	} else if c.Date != "" {
		return Cadence{}, invalidCadence("only a one-off schedule has a date.")
	}
	return out, nil
}

func uniqueLower(days []string) []string {
	var out []string
	for _, d := range days {
		d = strings.ToLower(strings.TrimSpace(d))
		if !slices.Contains(out, d) {
			out = append(out, d)
		}
	}
	return out
}

// allows says whether the cadence runs on a day of the week.
func (c Cadence) allows(day time.Weekday) bool {
	switch c.Repeat {
	case RepeatWeekdays:
		return day >= time.Monday && day <= time.Friday
	case RepeatWeekly:
		return slices.Contains(c.Days, weekdays[day])
	default:
		return true
	}
}

// at is the cadence's clock time on the local day `days` from t's.
func (c Cadence) at(t time.Time, days int, loc *time.Location) time.Time {
	clock, _ := time.Parse("15:04", c.Time)
	local := t.In(loc)
	return time.Date(local.Year(), local.Month(), local.Day()+days, clock.Hour(), clock.Minute(), 0, 0, loc)
}

// Next is the first time the cadence runs after t, if it ever will.
func (c Cadence) Next(t time.Time) (time.Time, bool) {
	loc, err := time.LoadLocation(c.Timezone)
	if err != nil {
		return time.Time{}, false
	}
	if c.Repeat == RepeatOnce {
		once, ok := c.once(loc)
		return once, ok && once.After(t)
	}
	for days := 0; days <= 8; days++ {
		if run := c.at(t, days, loc); run.After(t) && c.allows(run.Weekday()) {
			return run, true
		}
	}
	return time.Time{}, false
}

// Previous is the last time the cadence was due at or before t, if ever.
func (c Cadence) Previous(t time.Time) (time.Time, bool) {
	loc, err := time.LoadLocation(c.Timezone)
	if err != nil {
		return time.Time{}, false
	}
	if c.Repeat == RepeatOnce {
		once, ok := c.once(loc)
		return once, ok && !once.After(t)
	}
	for days := 0; days >= -8; days-- {
		if run := c.at(t, days, loc); !run.After(t) && c.allows(run.Weekday()) {
			return run, true
		}
	}
	return time.Time{}, false
}

func (c Cadence) once(loc *time.Location) (time.Time, bool) {
	day, err := time.Parse("2006-01-02", c.Date)
	if err != nil {
		return time.Time{}, false
	}
	clock, _ := time.Parse("15:04", c.Time)
	return time.Date(day.Year(), day.Month(), day.Day(), clock.Hour(), clock.Minute(), 0, 0, loc), true
}
